### PR TITLE
Tour Mode

### PR DESCRIPTION
**Pull request by Xuan-1998**
Current Changes in Tour Mode Choice
Please refer to original design of this PR to see more details: https://github.com/LBNL-UCB-STI/beam/pull/3604

- - -
This change is [Reviewable](https://reviewable.io/reviews/lbnl-ucb-sti/beam/3605)

### CODE REVIEW

**Comment by Xuan-1998** [comment]
image
@zneedell 
This is what's happening now when tour mode is CAR_BASED, people are using walk for pretty long trips with `personalVehicleAvailable = True`

**Comment by zneedell** [comment]
> image
> 
> @zneedell This is what's happening now when tour mode is CAR_BASED, people are using walk for pretty long trips with `personalVehicleAvailable = True`

Hmmm, not great. Looks to me like there are no alternatives available, so do you think that means this is an emergency walk trip? Perhaps either the route couldn't be constructed or someone took the vehicle before the person could use it? Are these both the first legs in tours or this person use a car earlier in the day?

Do long walk trips happen more in this branch than they did otherwise?

**Comment by Xuan-1998** [comment]
> > image
> > @zneedell This is what's happening now when tour mode is CAR_BASED, people are using walk for pretty long trips with `personalVehicleAvailable = True`
> 
> Hmmm, not great. Looks to me like there are no alternatives available, so do you think that means this is an emergency walk trip? Perhaps either the route couldn't be constructed or someone took the vehicle before the person could use it? Are these both the first legs in tours or this person use a car earlier in the day?
> 
> Do long walk trips happen more in this branch than they did otherwise?

Yes, I think it is. This person uses walk as the trip mode for the whole tour mode but the personal vehicle is always available, which means the route couldn't be constructed 
image

Yes, the long walk trips happen more in this branch than they did otherwise

**Comment by zneedell** [comment]
Can you tell if this person needed to replan at all first? If so, it might be that we need to clear that person's tour mode here: https://github.com/LBNL-UCB-STI/beam/blob/0624c2be27fde5d78d255a4142e1a9bddda4455f/src/main/scala/beam/agentsim/agents/modalbehaviors/ChoosesMode.scala#L1355, at least if it's the first trip of their tour

Perhaps we could change what is printed in the mode choice event here https://github.com/LBNL-UCB-STI/beam/blob/0624c2be27fde5d78d255a4142e1a9bddda4455f/src/main/scala/beam/agentsim/agents/modalbehaviors/ChoosesMode.scala#L1299 to give `filteredItinerariesForChoice` rather than `itinerariesOfCorrectMode` to see if other possible itineraries are being filtered out unnecessarily

**Comment by Xuan-1998** [comment]
> Can you tell if this person needed to replan at all first? If so, it might be that we need to clear that person's tour mode here:
> 
> https://github.com/LBNL-UCB-STI/beam/blob/0624c2be27fde5d78d255a4142e1a9bddda4455f/src/main/scala/beam/agentsim/agents/modalbehaviors/ChoosesMode.scala#L1355
> 
> , at least if it's the first trip of their tour
> Perhaps we could change what is printed in the mode choice event here
> 
> https://github.com/LBNL-UCB-STI/beam/blob/0624c2be27fde5d78d255a4142e1a9bddda4455f/src/main/scala/beam/agentsim/agents/modalbehaviors/ChoosesMode.scala#L1299
> 
> to give `filteredItinerariesForChoice` rather than `itinerariesOfCorrectMode` to see if other possible itineraries are being filtered out unnecessarily

I don't think it's the reason, the only replanning events are happening because of the resource unavailable of ride-hail and missing transit
image
And I think it's neither a network issue nor a routing problem because we didn't see a lot of long walkers in Dima's version and the mode split is changing normally. I tried a run of 30% population on current version, it can't even finished because of the following errors:
1. 16:47:38.406 [main] ERROR b.s.p.DefaultPopulationAdjustment - Not all agents have person attributes - is attributes file missing ?
2. 16:52:36.462 [ClusterSystem-akka.actor.default-dispatcher-7] ERROR b.a.i.parking.ParkingZoneFileUtils$ - [1] Failed to match row of parking configuration '{locationY=37.78637527, pricingModel=FlatFee, parkingType=Public, numStalls=35, chargingType=NoCharger, reservedFor=Any, feeInCents=0, locationX=-122.4049044, parkingZoneId=3, taz=12}' to expected schema
16:52:36.462 [ClusterSystem-akka.actor.default-dispatcher-7] ERROR b.a.i.parking.ParkingZoneFileUtils$ - [2] Failed to match row of parking configuration '{locationY=37.78487208, pricingModel=FlatFee, parkingType=Public, numStalls=19, chargingType=NoCharger, reservedFor=Any, feeInCents=0, locationX=-122.4008757, parkingZoneId=284, taz=12}' to expected schema
16:52:36.462 [ClusterSystem-akka.actor.default-dispatcher-7] ERROR b.a.i.parking.ParkingZoneFileUtils$ - [4] Failed to match row of parking configuration '{locationY=37.785876, pricingModel=FlatFee, parkingType=Public, numStalls=35, chargingType=NoCharger, reservedFor=Any, feeInCents=0, locationX=-122.4089225, parkingZoneId=4, taz=7}' to expected schema
16:52:36.463 [ClusterSystem-akka.actor.default-dispatcher-7] ERROR b.a.i.parking.ParkingZoneFileUtils$ - [8] Failed to match row of parking configuration '{locationY=37.7839879, pricingModel=FlatFee, parkingType=Public, numStalls=23, chargingType=NoCharger, reservedFor=Any, feeInCents=0, locationX=-122.412408, parkingZoneId=34, taz=8}' to expected schema
16:52:36.463 [ClusterSystem-akka.actor.default-dispatcher-7] ERROR b.a.i.parking.ParkingZoneFileUtils$ - [16] Failed to match row of parking configuration '{locationY=37.8005161, pricingModel=FlatFee, parkingType=Public, numStalls=27, chargingType=NoCharger, reservedFor=Any, feeInCents=0, locationX=-122.2720799, parkingZoneId=162, taz=968}' to expected schema
16:52:36.463 [ClusterSystem-akka.actor.default-dispatcher-7] ERROR b.a.i.parking.ParkingZoneFileUtils$ - [32] Failed to match row of parking configuration '{locationY=37.793177, pricingModel=FlatFee, parkingType=Public, numStalls=32, chargingType=NoCharger, reservedFor=Any, feeInCents=0, locationX=-122.398823, parkingZoneId=533, taz=1}' to expected schema
16:52:36.463 [ClusterSystem-akka.actor.default-dispatcher-7] ERROR b.a.i.parking.ParkingZoneFileUtils$ - [64] Failed to match row of parking configuration '{locationY=37.7811944, pricingModel=FlatFee, parkingType=Public, numStalls=31, chargingType=NoCharger, reservedFor=Any, feeInCents=0, locationX=-122.3999862, parkingZoneId=47, taz=21}' to expected schema
16:52:36.464 [ClusterSystem-akka.actor.default-dispatcher-7] ERROR b.a.i.parking.ParkingZoneFileUtils$ - [128] Failed to match row of parking configuration '{locationY=37.7630152, pricingModel=FlatFee, parkingType=Public, numStalls=23, chargingType=NoCharger, reservedFor=Any, feeInCents=0, locationX=-122.4264968, parkingZoneId=107, taz=104}' to expected schema
16:52:36.465 [ClusterSystem-akka.actor.default-dispatcher-7] ERROR b.a.i.parking.ParkingZoneFileUtils$ - [256] Failed to match row of parking configuration '{locationY=37.781123, pricingModel=FlatFee, parkingType=Public, numStalls=15, chargingType=NoCharger, reservedFor=Any, feeInCents=0, locationX=-122.2329915, parkingZoneId=226, taz=926}' to expected schema
3. 17:47:18.462 [ClusterSystem-akka.actor.default-dispatcher-45] ERROR b.a.a.h.HouseholdFleetManager - [1] No handler for BeamAgentSchedulerTimer
4. 17:54:45.648 [ClusterSystem-akka.actor.default-dispatcher-6] ERROR b.a.a.h.HouseholdFleetManager - [1] No handler for Failure(akka.pattern.AskTimeoutException: Recipient [Actor[akka://ClusterSystem/user/BeamMobsim.iteration/ParkingNetworkManager#730536009]] had already been terminated. Message of type [beam.agentsim.infrastructure.ParkingInquiry] was sent by [Actor[akka://ClusterSystem/user/BeamMobsim.iteration/population/1489591/Car#63500638]].)

**Comment by zneedell** [comment]
Were errors like this coming up in the runs with long walk trips yesterday? A problem with the parking manager could definitely lead to people not being able to find their cars.

**Comment by Xuan-1998** [comment]
No, these errors only show up in the bigger sample runs (30% one I launched this morning)

**Comment by zneedell** [comment]
Hmm, were there any parking related errors in the smaller runs? It might be unrelated but I think we should try to figure out what is causing the errors first in case it is leading to some of the other odd behavior.

The only other idea I can think of to help with debugging is recording all (not just filtered) modes here:
https://github.com/LBNL-UCB-STI/beam/blob/0624c2be27fde5d78d255a4142e1a9bddda4455f/src/main/scala/beam/agentsim/agents/modalbehaviors/ChoosesMode.scala#L1299

**Comment by Xuan-1998** [comment]
No there are no parking-related errors in the smaller runs. Thanks! Let me debugging on this

File: docs/inputs.rst [hunk 1]
```diff
@@ -238,17 +238,17 @@ Routing Configuration
...
+    mNetBuilder.toCRS = "EPSG:26910"    # UTM10N
+    travelTimeNoiseFraction = 0.0
+    maxDistanceLimitByModeInMeters {
+      bike = 40000
+    }
+    bikeLaneScaleFactor = 1.0
+    bikeLaneLinkIdsFilePath = ""
+  }
+  startingIterationForTravelTimesMSA = 0
+  overrideNetworkTravelTimesUsingSkims = false
```
Comment (dimaopen) [inline docs/inputs.rst]: It seems that the spacing was broken.

**Review by dimaopen — COMMENTED** [review]
Looks good. I would test Freight scenarios in order to see it works correctly.

**Comment by Xuan-1998** [comment]
> Hmm, were there any parking related errors in the smaller runs? It might be unrelated but I think we should try to figure out what is causing the errors first in case it is leading to some of the other odd behavior.
> 
> The only other idea I can think of to help with debugging is recording all (not just filtered) modes here:
> 
> https://github.com/LBNL-UCB-STI/beam/blob/0624c2be27fde5d78d255a4142e1a9bddda4455f/src/main/scala/beam/agentsim/agents/modalbehaviors/ChoosesMode.scala#L1299

image
Something is wrong here. Bike is showing up as an alternative for car_based trip inside `filteredItinerariesForChoice`

**Comment by Xuan-1998** [comment]
image
@dimaopen I remember you mentioned this before, you are seeing agents bring their cars out without bringing them back? I thought this is already solved. But It looks like this person's last trip end up walking. And I received the following warning:
09:33:27.737 WARN  beam.agentsim.agents.PersonAgent - [1] 597751 has received Finish while in state: WaitingToDrive, personId: 597751

**Comment by zneedell** [comment]
Hmmmm, yes this definitely sounds familiar, although I don't remember off the top of my head what the problem/solution was. But yeah it definitely looks like this person is forgetting about their car in some way or another after getting to work.

**Comment by Xuan-1998** [comment]
Test!

**Comment by nikolayilyin** [comment]
test!

**Review by dimaopen — COMMENTED** [review]
I haven't find any issues. But I think that we need to wait for @zneedell's approval.

**Comment by Xuan-1998** [comment]
Hi @dimaopen, if everything looks right to you. Do you have any idea this run failed: https://s3.us-east-2.amazonaws.com/beam-outputs/index.html#output/sfbay/gemini-base-2035__2022-09-24_14-39-04_ccw/

**Comment by dimaopen** [comment]
> Hi @dimaopen, if everything looks right to you. Do you have any idea this run failed: https://s3.us-east-2.amazonaws.com/beam-outputs/index.html#output/sfbay/gemini-base-2035__2022-09-24_14-39-04_ccw/

I didn't say that everything is right. I just didn't find any issues. What about this run I think that the reason is that [this `match`](https://github.com/LBNL-UCB-STI/beam/blob/d4bd9b1051c56560ace47a4e0fdcc42573f03c36/src/main/scala/beam/agentsim/agents/modalbehaviors/ChoosesMode.scala#L383) doesn't cover all possible values.

**Comment by Xuan-1998** [comment]
> > Hi @dimaopen, if everything looks right to you. Do you have any idea this run failed: https://s3.us-east-2.amazonaws.com/beam-outputs/index.html#output/sfbay/gemini-base-2035__2022-09-24_14-39-04_ccw/
> 
> I didn't say that everything is right. I just didn't find any issues. What about this run I think that the reason is that [this `match`](https://github.com/LBNL-UCB-STI/beam/blob/d4bd9b1051c56560ace47a4e0fdcc42573f03c36/src/main/scala/beam/agentsim/agents/modalbehaviors/ChoosesMode.scala#L383) doesn't cover all possible values.

Oh yeah! I see, thank you very much! I thought I already corrected this one.

**Comment by Xuan-1998** [comment]
@nikolayilyin Could you take a look at the error caused by this line: https://github.com/LBNL-UCB-STI/beam/blob/3ce8a3c461c2bbea604b68b3cb53c19d6e80e6a0/src/main/scala/beam/agentsim/agents/household/FastHouseholdCAVScheduling.scala#L432

**Comment by Xuan-1998** [comment]
> @nikolayilyin Could you take a look at the error caused by this line:
> 
> https://github.com/LBNL-UCB-STI/beam/blob/3ce8a3c461c2bbea604b68b3cb53c19d6e80e6a0/src/main/scala/beam/agentsim/agents/household/FastHouseholdCAVScheduling.scala#L432

solved @nikolayilyin

**Comment by Xuan-1998** [comment]
Test!

**Comment by Xuan-1998** [comment]
Test!

**Comment by Xuan-1998** [comment]
@nikolayilyin I merged it to develop with no conflicts but the testings are not all passing. Could you take a look at: https://beam-ci.tk/job/beam-forced-prs/4009/console

**Comment by Xuan-1998** [comment]
Test!

**Comment by Xuan-1998** [comment]
Test!

**Comment by HoneyTauOverTwo** [comment]
Test!

**Comment by Xuan-1998** [comment]
Test!

**Comment by dimaopen** [comment]
Test!

**Comment by Xuan-1998** [comment]
Test!

**Comment by Xuan-1998** [comment]
Test!

**Comment by zneedell** [comment]
Test!

**Comment by Xuan-1998** [comment]
Test!

**Comment by Xuan-1998** [comment]
Test!

**Comment by Xuan-1998** [comment]
test!

**Comment by Xuan-1998** [comment]
Test!

**Comment by zneedell** [comment]
Test!

**Comment by zneedell** [comment]
Test!

**Comment by Xuan-1998** [comment]
Test!

**Comment by Xuan-1998** [comment]
Test!

**Pull request closed without merging**